File: code/language/src/main/java/org/fbme/ide/iec61499/repository/PlatformElement.java
package org.fbme.ide.iec61499.repository;

import jetbrains.mps.smodel.CopyUtil;
import org.fbme.lib.common.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.mps.openapi.model.SNode;

public interface PlatformElement extends Element {

  @NotNull
  SNode getNode();

  @NotNull
  PlatformElementsOwner getOwner();

  @Override
  @NotNull
  default Element copy() {
    return getOwner().getAdapter(CopyUtil.copy(getNode()), getClass());
  }
}

File: code/language/src/main/kotlin/org/fbme/ide/iec61499/repository/PlatformElement.kt
package org.fbme.ide.iec61499.repository

import jetbrains.mps.smodel.CopyUtil
import org.fbme.lib.common.Element
import org.jetbrains.mps.openapi.model.SNode

interface PlatformElement : Element {
    val node: SNode
    val owner: PlatformElementsOwner
    override fun copy(): Element {
        return owner.getAdapter(CopyUtil.copy(node), javaClass)!!
    }
}
